### PR TITLE
Provide API for greater control of FocusGuide `autoFocus` logic

### PR DESCRIPTION
**Pull request by robik**
## Summary:

This PR implements additional API that is available from JavaScript code that provides more control over native FocusGuide `autoFocus` logic.

### Motivation

Focus management is crucial part in all TV applications. It is important to make it flexible enough that it covers many needs and cases of different applications. However, we stumbled a bit with current state of `<FocusGuide>` (which is a godsend, nevertheless).

https://github.com/react-native-tvos/react-native-tvos/assets/681837/60673ec7-adaf-4440-b566-4c484cfc1a45

As you can see in the video above, changing the tag updates contents of list of tiles (e.g. movies). When user changes the tag, the tile list is updated and the first tile should be focused when user presses down.
However, when the list of tiles is updated in place (without re-mounting) the last focused element is still tracked. You have remount the list for it to work.

### Solution

To achieve that, we are proposing a new `<FocusGuide>` prop named `entryMode` which specifies a strategy that decides which element should be focused upon entering the FocusGuide.

Entry mode requires `autoFocus` property to be set.

#### Available strategies

`entryMode` accepts one of two strategies which are available at the moment:

| Entry mode | Description                                                                                                                                                                                   |
|------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `restore`  | Behaves like current `<FocusGuide autoFocus>`. <br/> The first time focus enters the guide, the first visible element is focused, <br/> after re-entering the last focused element is focused (restored). |
| `first`    | When focus enters or re-enters the `focusGuide`, the first visible element is always focused.                                                                                                 |

New strategies can be added in the future as needed.

### New imperative methods

Additionally, for `restore` strategy, we propose two new imperative methods available on `FocusGuide` ref:

 - `resetLastFocus()` - allows to reset last focused element to first visible element imperatively
 - `updateLastFocus(target)` - imperatively sets specified `target` view as last focused element 

We believe that proposed changes allow for more flexibility when managing focus in TV applications.

### Alternative solutions

While the problem could be partially solved using `destinations`, it works in exclusion with `autoFocus`. This results in `<FocusGuide>` logic duality issue where two exclusive functionalities are provided. Our solution does not require developers to jump between them.

## Changelog:

```
- [General] [Added] - add `entryMode` prop to `FocusGuide`
- [General] [Added] - add `updateLastFocus` and `resetLastFocus` imperative methods to `FocusGuide` ref object
```

## Test Plan:

 - Updated rn-tester example

#### Resetting last focused element imperatively

```js
const [selectedTag, setSelectedTag] = React.useState(0);
const listGuideRef = React.useRef(null);

React.useEffect(() => {
  listGuideRef.current?.resetLastFocus();
}, [selectedTag]);

return (
  // .. tags
  // list of tiles:
  <TVFocusGuide autoFocus ref={listGuideRef}>
    ...
  </TVFocusGuide>
);
```

https://github.com/react-native-tvos/react-native-tvos/assets/681837/a8c6e0e3-5503-428b-a4d5-a59037a2fd17

#### List of elements with `entryMode="first"`

https://github.com/react-native-tvos/react-native-tvos/assets/681837/d3b96883-2a28-4f5a-9180-43aa390293f7

### CODE REVIEW

File: packages/react-native/Libraries/Components/TV/TVFocusGuideView.js [hunk 1]
```diff
@@ -78,6 +87,22 @@ function TVFocusGuideView(
     [],
   );
 
+  const updateLastFocus = React.useCallback((target) => {
+    if (focusGuideRef.current) {
+      // Handle ref to component argument
+      const handle = (typeof target === "number") ? target : findNodeHandle(target)
+      if (handle) {
+        Commands.updateLastFocus(focusGuideRef.current, target);
```
[the remark this answers is not in the document]
Comment (robik) [reply packages/react-native/Libraries/Components/TV/TVFocusGuideView.js]: Right, looks like it slipped on rename. Fixed!

**Review by douglowder — CHANGES_REQUESTED** [review]
Thanks VERY much for this PR, we are looking at it carefully!

One missing piece I noticed: you modified `RCTTVView.m`, but did not make corresponding changes in the Fabric implementation for Apple TV. Most of your changes would go in `packages/react-native/React/Fabric/Mounting/ComponentViews/View/RCTViewComponentView.mm`, I think....

**Comment by robik** [comment]
Hey @douglowder

You are correct, I missed the changes for iOS fabric. This should be fixed now, as the missing code is added and tested on both TvOS and AndroidTV. 😅

As a note, this is my first contribution to `react-native` codebase, so let me know if anything else to be updated/changed!

**Review by yusufyildirim — CHANGES_REQUESTED** [review]
Thanks for the hard work! This has been one of the things I want to work on. I'll ask for some significant changes, though. Everything you want can be achieved without introducing a native change. Currently exposed APIs we have might seem a little restrictive you might think but I came up with the minimal API possible with maximum composibility in mind. So, properties like `destinations`, `autoFocus` work in harmony intead of overriding the behaviors of each other so you can compose them and come up with a new capability that is not directly exposed by the `TVFocusGuideView`. I'll give an example below.

1. I think we shouldn't expose a way to set the `lastFocusedElement`, it's kind of against it's purpose. It should be always be -literally- the last focused element. If you need focus to move to a specific location, we have `destinations` prop for that. I understand you want it to be focusable only "once" and delegate the rest to the `autoFocus` to keep track of the `lastFocusedElement` and so on. You can achieve that behavior by setting `destinations`, listening `onFocus` of your destination element and setting `destination` back to empty while `autoFocus` set to true for your container. Focus will be redirected to the `destinations` you provide initially because `destinations` prop has priority --when redirecting focus-- over other mechanisms.  `TVFocusGuide` will keep track of the `lastFocusedElement` on the background as long as `autoFocus` is true. When you empty `destinations`, `lastFocusedElement` will be the candidate for focus redirection the next time focus moves inside the container. Might sound a little work but you realistically need a way to keep track of your elements somewhere in a Context or something and automate this process by creating your own abstraction. We use such abstraction at work that I'm planning to open-source in the future but I don't know when.  
2. It makes sense to provide a way to reset the `lastFocusedElement`, though. I was aware of the list use-case you described above while implementing `autoFocus`. That's also already possible to achieve. Whenever you set `autoFocus` no matter the value, it resets the `lastFocusedElement`. So, a simple call like `tvFocusGuideRef.setNativeProps({ autoFocus: false /*or true depending on what's the current autoFocus value */})` achieves that without even triggering a react render. I'm in for exposing a new imperative call something like `resetLastFocusedElement` with this implementation.
4. `entryMode` is also technically achievable I believe if you set an invalid ID in `destinations` since the focus restoration logic will kick in in that case and restore the focus to the first focusable element every time. I accept it's a bit too hacky, though. 😄 I was thinking of introducing a way to override the automatic `lastFocusedElement` tracking functionality for this use-case. So, I thought we can introduce a new prop called   `rememberLastFocusedElement` which would disable `lastFocusedItem` tracking on the native side (when provided false) with a couple of lines of code and the focus recovery/redirection logic handles the rest.

I'm trying to keep native changes and new API surface as little as possible as you can tell and this is solely for maintainability. Let's discuss!

**Comment by robik** [comment]
Thanks for your insightful feedback!

You raised some really good points in your commend, I'll try my best to answer and cover all points you made below 🙂

> Everything you want can be achieved without introducing a native change.

This is indeed true. However, I think that focus is one of those areas that works better when handled on native side. Managing focus logic in react tree is extremely verbose, hard to maintain and error-prone. Having this logic in native code yields a bit better performance. 

In my opinion, this is something that should be done in different layer/model, past react tree, which can be mutated programmatically, just as it is done in DOM on web platform. Without this, the focus logic simply leaks through your components and contexts when it shouldn't, you have to pass `ref`s all around, even drill them deep in rare cases. This is only my subjective opinion through.

I am not saying it must be added in the existing `FocusGuide` component though, I'm okay with it implemented as a separate component built on top of `FocusGuide`.

> Currently exposed APIs we have might seem a little restrictive you might think but I came up with the minimal API possible with maximum composibility in mind. 

I agree. I think proposed additions do not make it _less minimal_ API, but more complete in its current boundaries. I mean, we can also do tracking of last focused element and re-focusing on the JS side (with a more low-level API to make it happen), removing the need for `autoFocus`.

> So, properties like destinations, autoFocus work in harmony intead of overriding the behaviors of each other so you can compose them and come up with a new capability that is not directly exposed by the TVFocusGuideView.

They are a powerful weapons, true. The problem with this approach is that they do not express intent clearly enough. Its kind of like arguing we don't need `Array.filter/map/reduce` as we already have very powerful, albeit more generic `Array.forEach`.

> 1. I think we shouldn't expose a way to set the lastFocusedElement, it's kind of against it's purpose. It should be always be -literally- the last focused element. 

My idea is that if we have some logic on the native side, and we can easily make it more flexible, why shouldn't we? This simple addition may allow developers to implement some convoluted cases if needed, without workarounds that are impossible to decipher, and all at low cost (subjectively). 
If you look at this as more modular/parametrized approach, this is simply a piece of logic that can be altered/parametrized, like a function or class. If you want to override currently tracked lastElement why should be block user from doing so? This might be a really rare case, but if the user needs it, should they really work around the solution or re-implement it?

> 1. [...] You can achieve that behavior by setting destinations, listening onFocus of your destination element and setting destination back to empty while autoFocus set to true for your container. Focus will be redirected to the destinations you provide initially because destinations prop has priority --when redirecting focus-- over other mechanisms. TVFocusGuide will keep track of the lastFocusedElement on the background as long as autoFocus is true.

This is possible to write, but a pain to read and maintain. The process you described leverages implementation detail, which may be guaranteed by the library, or just be an result of current implementation. This comes back to expressing intent I was referring to. Personally, if I would find such code, I'd think the code is fighting with the library, rather than using it correctly.

> 2. It makes sense to provide a way to reset the lastFocusedElement, though. I was aware of the list use-case you described above while implementing autoFocus. [..] I'm in for exposing a new imperative call something like resetLastFocusedElement with this implementation.

Glad to hear this! 🙏 

> 3. entryMode is also technically achievable I believe if you set an invalid ID in destinations since the focus restoration logic will kick in in that case and restore the focus to the first focusable element every time.

I was thinking of posting a suggestion in the future to merge these two. I might not see all the cases with `destinations` prop that are currently implemented, so feel free to correct me if I am wrong on this one, but I think that `entryMode` can be seen as a superset of current solution. We could make current `destinations` prop, which takes priority over `autoFocus` a separate `entryMode`, e.g. `entryMode="manual" destinations={[]}`. This way the exposed API of `<FocusGuide>` is more straightforward, and is more explicit in the change of component behavior. This would be a breaking change though.

Also, regarding the proposed alternative, my goal is to provide that API that is not only is flexible, but also expresses developer intent clearly.

> I'm trying to keep native changes and new API surface as little as possible as you can tell and this is solely for maintainability. Let's discuss!

I completely understand your point here! But the issue here is what should be handled by the library, and what should be done by library users. Should the library handle such cases, or only provide primitives?

As I mentioned on the beginning, I think because it better suited to be part of native code, it is better to have it in the library. If it would be something that can be easily done on JS, it could be a independent npm package.

Hope I covered all your points and concerns, WDYT?

**Comment by yusufyildirim** [comment]
> This is indeed true. However, I think that focus is one of those areas that works better when handled on native side. Managing focus logic in react tree is extremely verbose, hard to maintain and error-prone. Having this logic in native code yields a bit better performance.

Everything I described and proposed above works on the native side. The only exception is listening `onFocus` event which is also sent from the native. We'd be doomed if we don't even want to pay the cost of listening an event and calling an another native function as it's side effect. Javascript is fast enough for such case.

> In my opinion, this is something that should be done in different layer/model, past react tree, which can be mutated programmatically, just as it is done in DOM on web platform. Without this, the focus logic simply leaks through your components and contexts when it shouldn't, you have to pass refs all around, even drill them deep in rare cases. This is only my subjective opinion through.

We ask for refs only when we have to. There's no other way to identify which native element you're referring to. I don't like this as well but if you don't want to pass refs around, you need your own abstraction to solve this problem. It's very much possible. We can also provide this in the core, I think about it every day but it's not a simple operation. I personally try to give you the tools so you can build your own solutions as it fits your needs.

> I mean, we can also do tracking of last focused element and re-focusing on the JS side (with a more low-level API to make it happen), removing the need for autoFocus.

I think I failed to explain what I meant there. I'm trying to keep the native footprint low here, not the JS. It's perfectly fine to expose new functionality to the user on the JS land. It doesn't mean they won't be native-backed. I'm saying, all of what you've done so far are possible with the current native functionality. Even exposed to the user so they can come up with their own solutions if we fail to provide. That's how they're designed. So, let's provide features to the users without introducing new native code if possible.

> The problem with this approach is that they do not express intent clearly enough. Its kind of like arguing we don't need Array.filter/map/reduce as we already have very powerful, albeit more generic Array.forEach

The answer to your question lies directly inside it. Those are convenience features built on top of good-old  regular loops. It wouldn't be possible to build them if we didn't have loops in the first place. `map` etc. introduced with ES6 (2015), `forEach` introduced in ES5 (2009?). We were perfectly fine doing our jobs before them. Yeah, it was a lot more chore but it was possible because we had the regular for/while loops. People even built `lodash` and stuff for this purpose. I'm not comparing us with a programming language but I'm trying to tell we're in a similar process here. And nobody said we don't need convenience feature, just said "Everything you want can be achieved without introducing a native change. "

> My idea is that if we have some logic on the native side, and we can easily make it more flexible, why shouldn't we?

Because it's not how interface design works. We wouldn't need the `private`/`public` concepts in programming if we were exposing everything to the user just because we can.

> If you want to override currently tracked lastElement why should be block user from doing so?

Because it's is how interface design works, this time. We don't expose things to the users that we don't want them to play around with. They can easily abuse it and shot themselves in the foot. Overriding `lastFocusedElement` is not something that should be done. Its name very clearly indicates what its responsibility is. You want control over it because you want to redirect the focus to an another element when focusing in to the container. So you're already ready to abuse it 😄  You're trying to solve the problem in the wrong place, what you want is not its responsibility.

>  Personally, if I would find such code, I'd think the code is fighting with the library, rather than using it correctly.

These are the features we provide to the user. It's perfectly fine to use them together like lego pieces. `initialDestination` is a feature we use at work (implemented ourselves with the recipe I shared above) incredibly extensively, it's amazing. It's not super simple to implement in the core due to various reasons, though. I'm working on something that'll make it relatively easy to build, though.

> My concern on this is how much of the process you described is guaranteed by the library, and how much of it leverages implementation detail
It's written in the readme if you check the `autoFocus` description. And it should also be documented in the types. The wording can be improved, though. Simply speaking, those are the features that we guarantee. I wouldn't recommend you building something off of an implementation detail. 

> is it really more minimal? We have two functions/props that have multiple effects you can use to do various stuff. 

Understanding what `FocusGuide` is in tvOS helps understand why `TVFocusGuideView` built in such way. It's basically tvOS's `FocusGuide` on steroids. It's also the cross platform version of it. That's why we have `destinations` prop for example. I'd recommend this article: https://www.enekoalonso.com/articles/understanding-and-visualizing-uifocusguide-on-tvos

React Native -unlike Flutter-, doesn't try to provide the users everything they need. It mostly provides the basics and lets the community build the rest adjusted to their needs, either completely from scratch or on top of the existing primitives. So it's not a goal to solve every single use-case. The goal is to provide the infra for them. That's how I see it, at least. Some folks may disagree, I respect that.

We'll keep providing new convenience features to the users, don't get me wrong. I have a list of focus related improvements waiting for me to implement. They'll provide incredible value to the users. I'll just personally try to write and maintain minimal amount of native code _if possible_. Which isn't, in most cases, but it mostly is for the features you're implementing here.

**Comment by robik** [comment]
>  Everything I described and proposed above works on the native side. The only exception is listening onFocus event which is also sent from the native. We'd be doomed if we don't even want to pay the cost of listening an event and calling an another native function as it's side effect. Javascript is fast enough for such case.

I agree its good to do the event handing on the JS side, its part of the application logic after all. The issue I have with this, and it is mostly related to working with focus, is that you need to drill and poison your UI tree with `ref`s across various components. This is getting quickly overwhelming when some conditions are involved. I believe that React code is at its best when most parts are done declaratively, falling back to imperative calls only when needed. I elaborate on this a bit more below.

> I'm trying to keep the native footprint low here, not the JS. It's perfectly fine to expose new functionality to the user on the JS land. It doesn't mean they won't be native-backed. I'm saying, all of what you've done so far are possible with the current native functionality. Even exposed to the user so they can come up with their own solutions if we fail to provide. That's how they're designed. So, let's provide features to the users without introducing new native code if possible.

I understand. In this case, perhaps it would be better to expose all the mechanisms, events and APIs needed to implement `autoFocus` on the JS side? So that it can be a JS component, users can create their own variants and upload them to npm, without the need to patch rn-tv 🤔

What I am worried about (and that's why I referred to DOM like tree) is that when reacting to a focus event, you cannot iterate through view tree (as in the parents, or the children recursively). You cannot simply add state or logic to generic views in React. You can change the native UI tree, but to do it in React is extremely verbose, because of the UI-as-code approach and the reactivity flow.
You may end up in a situation where you'd need to create custom View and re-implement all other components and libraries that use them. This is how crucial focus is.

> Because it's not how interface design works. We wouldn't need the private/public concepts in programming if we were exposing everything to the user just because we can.

True. But in this case, a piece of logic is done (autoFocus) is done on the native side without the possibility to overload it whatsoever. In case we'd be writing it all on native side, I'd just make it a `protected` method arguably, inherit the class (in case of OOP pattern) and overload the method to adjust the logic.

> Because it's is how interface design works, this time. We don't expose things to the users that we don't want them to play around with. They can easily abuse it and shot themselves in the foot. Overriding lastFocusedElement is not something that should be done. 

I'm okay with removing it 😁. For me it is a basic building block of `autoFocus`, which `resetLastFocus` method is a special case of `setLastFocus(null)`, exposed for clarity.

However, if we want to provide most basic blocks to the users on the native side, we should give users possibility to track focus changes within FocusGuide and then they could write their own `autoFocus` in the JS side. This would give them the possibility to alter the logic if needed. Its hard to do right now as it is a breaking change.

The issue I am trying to solve right now, is that the abstraction is bit leaky. Its impossible to recreate autoFocus on JS right now, or to alter existing one without changing native code. This is a violation of [Open/Closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle), while mainly used for OOP code, is a good rule to live by anyways.

> It's written in the readme if you check the autoFocus description. And it should also be documented in the types. The wording can be improved, though. Simply speaking, those are the features that we guarantee. I wouldn't recommend you building something off of an implementation detail.

Allrighty, just wanted to clarify whenever all of it is guaranteed (on AndroidTV as well). 😄

> Understanding what FocusGuide is in tvOS helps understand why TVFocusGuideView built in such way. It's basically tvOS's FocusGuide on steroids. It's also the cross platform version of it. That's why we have destinations prop for example. I'd recommend this article: https://www.enekoalonso.com/articles/understanding-and-visualizing-uifocusguide-on-tvos

I'm not really an expert in Apple Frameworks. But I'd be wary of designing close to FocusGuide. The API was designed for UIKit which leverates OOP heavily, React uses a different approach and data flow, so the API design should reflect that accordingly.

> React Native -unlike Flutter-, doesn't try to provide the users everything they need. It mostly provides the basics and lets the community build the rest adjusted to their needs, either completely from scratch or on top of the existing primitives. So it's not a goal to solve every single use-case. The goal is to provide the infra for them. That's how I see it, at least. Some folks may disagree, I respect that.

I see your point on this one. See my suggestion on new primitives for recreating autofocus on JS.

**Comment by yusufyildirim** [comment]
We discussed about some of it in the Expo `react-native-tv` channel but I also want to leave a comment here for posterity.

In summary, we approach to the problems from different perspectives. I'm trying to keep things lean while you ask for the maximum control possible. We can achieve flexibility in both ways, with different implementations. You ask for a raw control which is a no-no in my book. I'm on the fence of building safer features, keeping the balance when exposing stuff. Here's the summary of what I think about the things introduced in this PR:

**entryMode**: You want a way move focus to the first element all the time, which is fair. I'd love to introduce an API that'd give you a way you do this sort of stuff in an easier way, the closest we got is `destinations` prop. I don't want to provide a specific solution for this specific use-case. What's gonna happen if somebody comes up with a use-case that requires them to move focus to the third element all the time. Instead, I'll create a recipe how it can be achieved with the current primitives we have so everybody can build a solution for every similar use-case they have.

**resetLastFocus:** I get this feature could be useful but it can be a foot gun at the same time, depending on how you use it. I forgot to add one important thing above regarding this. The list example you have is against one of the most important rules of React, `key`s should be unique. You shouldn't cheat the renderer by providing the same `key` to an essentially different elements. You can easily find yourself in debugging some weird bugs especially if the components you render are stateful. I'd recommend reading React doc one more time: https://react.dev/learn/rendering-lists. Still, I know it could be useful to have this functionality, so we might introduce it, but with a slightly different implementation.

**updateLastFocus:** I'm probably repeating myself but this is totally against the purpose tracking last focused element. If we can change it, it's not last focused element anymore, hence the name doesn't make sense. You're looking for a solution at the wrong place. What you want instead is `initialDestination` which will let you define a redirection point only for the next (or initial) visit. You expect it to remember the `lastFocusedElement` (fallback to autoFocus) after that. We've been using such feature at work, we implemented it ourselves. I think it's time to move it to the core, I'll take care of it.

I'll close this PR since it's too big and we don't want the most of it. I'll build the necessary things myself and come up with `Recipes` in the wiki to clear things out for everybody. Thanks for the amazing work!

**Pull request closed without merging**